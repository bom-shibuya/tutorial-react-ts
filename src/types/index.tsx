export type SquareValue = 'O' | 'X' | null;
export type Squares = SquareValue[];