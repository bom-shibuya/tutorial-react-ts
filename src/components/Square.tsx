import * as React from 'react';
import { SquareValue } from '../types'

export interface ISquareProps {
  value: SquareValue;
  onClick: () => void;
}

export const Square = ({value, onClick}: ISquareProps) => (
  <button
    className="square"
    onClick={onClick}>
    {value}
  </button>
)
