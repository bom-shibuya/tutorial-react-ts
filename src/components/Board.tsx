import * as React from 'react';

import { Squares } from '../types';
import { Square } from './Square';

export interface IBoardProps {
  squares: Squares;
  handleClick: (i:number) => void;
}

export const Board = ({handleClick, squares}: IBoardProps) => {

  const renderSquare = (i: number) => {
    const onClick = () => handleClick(i)
    return (
      <Square
        value={squares[i]}
        onClick={onClick}/>
    )
  }

  return (
    <div>
      <div className="border-row">
        {renderSquare(0)}
        {renderSquare(1)}
        {renderSquare(2)}
      </div>
      <div className="border-row">
        {renderSquare(3)}
        {renderSquare(4)}
        {renderSquare(5)}
      </div>
      <div className="border-row">
        {renderSquare(6)}
        {renderSquare(7)}
        {renderSquare(8)}
      </div>
    </div>
  )
}
