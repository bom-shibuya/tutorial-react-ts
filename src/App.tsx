import * as React from 'react';
import './App.css';
import { Board } from './components/Board'
import { Squares } from './types'

export const calculateWinner = (squares: Squares) => {
  const lines = [
    [0, 1, 2],
    [3, 4, 5],
    [6, 7, 8],
    [0, 3, 6],
    [1, 4, 7],
    [2, 5, 8],
    [0, 4, 8],
    [2, 4, 6],
  ];

  for (const line of lines) {
    const [a, b, c] = line;
    if (squares[a] && squares[a] === squares[b] && squares[a] === squares[c]) {
      return squares[a]; // 'X' of 'O'
    }
  }
  return null;
}


export interface IGameState {
  currentHistory: number;
  squares: Squares;
  xIsNext: boolean;
  history: Array<{
    squares: Squares
  }>
}

class App extends React.Component<{}, IGameState> {
  public state: IGameState = {
    currentHistory: 0,
    history: [{
      squares: Array(9).fill(null)
    }],
    squares: Array(9).fill(null),
    xIsNext: true,
  }

  public handleClick = (i: number) => {
    const squares = this.state.squares.slice();
    if (calculateWinner(squares) || squares[i]) {
      return;
    }
    squares[i] = this.state.xIsNext ? 'X' : 'O';
    const history = [...this.state.history.slice(0, this.state.currentHistory + 1), {
      squares
    }];

    this.setState({
      currentHistory: this.state.currentHistory + 1,
      history,
      squares,
      xIsNext: !this.state.xIsNext,
    })
  }

  public handleHistory(i: number) {
    const selectedHistory = this.state.history[i].squares.slice();
    this.setState({
      currentHistory: i,
      squares: selectedHistory,
      xIsNext: (i % 2 === 0)
    })

  }

  public render() {
    const winner = calculateWinner(this.state.squares);
    const status = winner ? `Winner: ${winner}` : `Next Player: ${this.state.xIsNext ? 'X' : 'O'}`;
    const histories = this.state.history.map((history, i) => {
      const text = i === 0 ? 'Back to game start' : `Back to #${i}`;
      const handleHistory = () => this.handleHistory(i)
      return (
        <li key={i}>
          <button onClick={handleHistory}>{text}</button>
        </li>
      )
    })
    return (
      <div className="game">
      <div className="gema-board">
        <Board
          squares={this.state.squares}
          handleClick={this.handleClick}/>
      </div>
      <div className="game-info">
        <div>{ status }</div>
        <ol>
          {histories}
        </ol>
      </div>
    </div>
    )
  }
}
export default App;
