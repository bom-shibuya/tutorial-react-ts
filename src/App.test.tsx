import { mount, ReactWrapper, render, shallow, ShallowWrapper } from 'enzyme';
import * as React from 'react';
import * as ReactDOM from 'react-dom';
import App, { calculateWinner, IGameState } from './App';
import { Board } from './components/Board'
import { Square } from './components/Square'
import { Squares } from './types';

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<App />, div);
  ReactDOM.unmountComponentAtNode(div);
});

const noWin: Squares = [
  'X', 'O', 'X',
  null, 'O', null,
  'X', null, 'X',
];
const xWin: Squares = [
  'X', 'O', 'X',
  'X', 'O', null,
  'X', null, 'X',
];
const oWin: Squares = [
  'X', 'O', 'X',
  'X', 'O', null,
  null, 'O', 'X',
];

describe('calculateWinner', () => {
  test('勝負が決まってない場合はnullを返す', () => {
    expect(calculateWinner(noWin)).toBeNull();
  });

  test('Xが勝った場合はXを返す', () => {
    expect(calculateWinner(xWin)).toBe('X');
  });

  test('Oが勝った場合はOを返す', () => {
    expect(calculateWinner(oWin)).toBe('O');
  });
})


describe('App test', () => {

  let app: ShallowWrapper;
  let setStateSpy: jest.SpyInstance;

  beforeEach(() => {
    app = shallow(<App />);
    setStateSpy = jest.spyOn(App.prototype, 'setState');
  })

  test('子コンポーネントが存在する', () => {
    expect(app.find(Board).length).toBe(1);
  })

  describe('handleClickの呼び出し', () => {
    const clickNum = 0;
    const expectState: IGameState = {
      currentHistory: 1,
      history: [
        {
          squares: Array(9).fill(null)
        },
        {
          squares: [
            'X', ...Array(8).fill(null)
          ]
        }
      ],
      squares: [
        'X', ...Array(8).fill(null)
      ],
      xIsNext: false
    }

    test('任意のボタンをクリック', () => {
      app.instance().handleClick(clickNum);
      expect(setStateSpy).toHaveBeenCalledWith(expectState);
    })

    test('同じ場所はクリックできない', () => {
      app.instance().handleClick(clickNum);
      app.instance().handleClick(clickNum);
      expect(setStateSpy).toHaveBeenCalledWith(expectState);
    })

    test('結果が確定後はクリックできない', () => {
      const endState: IGameState = {
        currentHistory: 2,
        history: [
          ...expectState.history,
          {
            squares: xWin
          }
        ],
        squares: xWin,
        xIsNext: false
      };
      app.setState(endState);
      app.instance().handleClick(clickNum);
      expect(setStateSpy).toHaveBeenCalledWith(endState);
    })
  })

  describe('handleHistoryの呼び出し', () => {

    let appMount: ReactWrapper;
    let square1: ReactWrapper;
    let square2: ReactWrapper;
    let square3: ReactWrapper;

    beforeEach(() => {
      appMount = mount(<App />);
      square1 = appMount.find(Square).at(0);
      square2 = appMount.find(Square).at(1);
      square3 = appMount.find(Square).at(2);
    })

    test('任意の位置に戻れる', () => {
      square1.simulate('click');
      const stashState = {...appMount.instance().state};
      square2.simulate('click');
      square3.simulate('click');

      const historyBtn2 = appMount.find('li').at(1).find('button');
      historyBtn2.simulate('click');
      expect(setStateSpy).toHaveBeenCalledWith(stashState);
      // expect()
    })

    test('任意の位置に戻った後squareがクリックされるとhistoryが更新される', () => {
      square1.simulate('click');
      square2.simulate('click');
      square3.simulate('click');
      const historyBtn2 = appMount.find('li').at(1).find('button');
      historyBtn2.simulate('click');
      // 戻っただけではヒストリーは更新されない
      expect(appMount.instance().state.history.length).toBe(4);
      // 戻った後、squareが押されると以降の部分が更新される
      square3.simulate('click');
      expect(appMount.instance().state.history.length).toBe(3);
    })
  })



})
