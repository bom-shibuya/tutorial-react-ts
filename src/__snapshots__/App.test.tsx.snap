// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<div
  className="game"
>
  <div
    className="gema-board"
  >
    <div>
      <div
        className="border-row"
      >
        <button
          className="square"
          onClick={[Function]}
        />
        <button
          className="square"
          onClick={[Function]}
        />
        <button
          className="square"
          onClick={[Function]}
        />
      </div>
      <div
        className="border-row"
      >
        <button
          className="square"
          onClick={[Function]}
        />
        <button
          className="square"
          onClick={[Function]}
        />
        <button
          className="square"
          onClick={[Function]}
        />
      </div>
      <div
        className="border-row"
      >
        <button
          className="square"
          onClick={[Function]}
        />
        <button
          className="square"
          onClick={[Function]}
        />
        <button
          className="square"
          onClick={[Function]}
        />
      </div>
    </div>
  </div>
  <div
    className="game-info"
  >
    <div>
      Next Player: X
    </div>
    <ol>
      <li>
        <button
          onClick={[Function]}
        >
          Back to game start
        </button>
      </li>
    </ol>
  </div>
</div>
`;
